include/swift/AST/Requirement.h

```diff
@@ -124,6 +124,55 @@ struct StructuralRequirement {
   bool inferred = false;
 };
 
+/// Represents an invalid requirement, such as `T: Int`.
+///
+/// Invalid requirements are recorded while computing the
+/// generic signature of a declaration, and diagnosed via
+/// \c diagnoseRequirementErrors .
+struct RequirementError {
+  /// The kind of requirement error.
+  enum class Kind {
+    InvalidConformance,
+    ConcreteTypeMismatch,
+    NonTypeParameter,
+    RedundantRequirement,
+  } kind;
+
+  /// The invalid requirement.
+  Requirement requirement;
+
+  SourceLoc loc;
+
+private:
+  RequirementError(Kind kind, Requirement requirement, SourceLoc loc)
+    : kind(kind), requirement(requirement), loc(loc) {}
+
+public:
+  static RequirementError forInvalidConformance(Type subjectType,
+                                                Type constraint,
+                                                SourceLoc loc) {
+    Requirement requirement(RequirementKind::Conformance, subjectType, constraint);
+    return {Kind::InvalidConformance, requirement, loc};
+  }
+
+  static RequirementError forConcreteTypeMismatch(Type type1,
+                                                  Type type2,
+                                                  SourceLoc loc) {
+    Requirement requirement(RequirementKind::SameType, type1, type2);
+    return {Kind::ConcreteTypeMismatch, requirement, loc};
+  }
+
+  static RequirementError forNonTypeParameter(Requirement req,
```

Comment:
forConflictingRequirement() perhaps, since in reality this means the concrete type doesn't satisfy the requirement